### ISSUE TITLE
Clarification on double-entry accounting and debit/credit

### ISSUE THREAD

**Issue by qskwood** [firefly-iii/firefly-iii#9259]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)
- [X] My request is not listed as [a very good idea, but unfortunately...](https://docs.firefly-iii.org/explanation/more-information/what-its-not/)
- [X] I've used [the search](https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue) and this has not been requested before.

### Description

Re: #3385 which was responded to with "No, sorry, not going to happen."

I was originally interested in this because it says on the website and GitHub repo

> Firefly III features a double-entry bookkeeping system.

> A [double-entry](https://en.wikipedia.org/wiki/Double-entry_bookkeeping_system) bookkeeping system.

Is the response from 4.5 years ago still accurate, and if so why? Coming from a financial accounting and double-entry bookkeeping background, the way Firefly III forces everything to be a withdrawal-deposit pair is less intuitive. For example, I searched for issues about this because I noticed it's impossible to have a negative value on an expense or revenue.

If I buy something on Amazon and return it the refund isn't revenue, it's a reduction to the expense account. Likewise if I receive a check for more than I earned the refund isn't an expense, it's a reduction to the revenue account. These are two very simple, common tasks for personal finance that the software can't handle because it's foundationally performing withdrawal-deposit pairs and not debits and credits.

In the same theme a paycheck doesn't reduce the value of a revenue account and increase the value of an asset account, it increases the value of both. No transfer between accounts is occurring, that's a single-entry accounting concept. It's not really correct to call this a double-entry accounting application when it not only currently does single-entry accounting but simply will not allow double-entry accounting.

I'd like to see the application UI get away from withdrawal-deposit and move to debit-credit:

- transactions that involve earning money should increase both revenue and asset accounts
  - this is a request for a UI change, not a backend math change
  - ![Screenshot 2024-09-22 151257](https://github.com/user-attachments/assets/1ecefa0d-270d-406e-8fa0-c94d74543402)
- transactions that involve spending money should increase expense and reduce asset accounts
  - this is a request for a UI change, not a backend math change
  - ![Screenshot 2024-09-22 151608](https://github.com/user-attachments/assets/74885210-3439-4e27-a615-8f97c3e1f609)
- transactions against revenue and expense accounts should allow negative numbers
  - this is a request for an input validation and UI change, not a backend math change
  - ![Screenshot 2024-09-22 151946](https://github.com/user-attachments/assets/74d71b53-f0bb-4380-9173-e12db5ea03a0)
  - ![Screenshot 2024-09-22 152302](https://github.com/user-attachments/assets/155e0044-cace-4ded-a114-53251747bf91)

### Solution

The application would do what it says on the tin, double-entry accounting. The following transactions would be possible:

- returning something to the store
- refunding a payment
- quickly understanding the transaction list under a revenue or expense to see examples of the above

I've only been using the software for a few days, so there could be other things it can't do that I haven't found.

### Additional context

Most of the changes necessary to fix this aren't probably in the backend, but in the way users interact with the application. The way things are calculated, etc. seems perfectly fine. For example, allowing a negative value to be plugged into an expense or revenue isn't (I assume) a change to backend logic but to frontend input validation that is, in my opinion, misplaced.

The rest is a matter of moving away from UIs that describe single-entry accounting processes to ones that are agnostic (to avoid confusing people that don't understand double-entry accounting) or are specifically related to double-entry accounting (since the application is described as using double-entry accounting anyway.)

**Comment by qskwood** [comment]
I forgot to annotate, but the idea of source and destination accounts in my last two screenshots is, in itself, an artifact of single-entry accounting and doesn't make any sense in double-entry accounting.

**Comment by JC5** [comment]
Hi, thanks for opening an issue. 

I'm aware of all of these inconsistencies. I designed these concepts when I started Firefly III over ten years ago after googling for five minutes how accounting works. 

Like I said last time, I'm not going to change things. I appreciate the feedback and it's absolutely required to do things better, but I have a billion other things on my list and conceptually the current WoW works for most people and for me. 

I will take this into my notes and pick things up for the new layout but I'm not going to change the fundamentals of Firefly III. I know you mentioned it's superficial. 

Thanks again. For an app that does this way better, see GnuCash.

**Comment by qskwood** [comment]
Thank you. If you could allow negative values in expenses and revenues, that might address the two things I need without fundamentally modifying how the software behaves.

**Comment by ikotamsl** [comment]
Dear James @JC5 
I was on my way to open the issue, that discusses almost the same topic, but I stumbled around this one. One thing that I want to add is that sometimes Expense accounts make no sense. I am not insisting on adding a double-entry, but it sometimes very confusing when you don't see your Expense account going down.

As an example, mentioned above, the return of a good is not a revenue, but a decrease on an expense account. If I do transactions like Asset -> Expense1 and then Expense1 -> Asset, in interface I will only see the result of the first transaction. Therefore, the balance of the account will be incorrect, while the transaction itself exists. Is it possible to at least render all the transactions and their balance. The feature may be hidden by the config of the server, because i realize that many people do not use their expense accounts as a traditional "accounting" account.

Thank you for your work and wish you all the best!

**Comment by jaredchurch** [comment]
Hi,

Adding another voice here, I'd love to see the ability to have a negative value on deposit/withdrawal transactions.

Based on assumption that this is php code, I'd be happy to take a stab at code adjustment for this (it has been a little while since I last worked with php, but that's ok). At the same time I don't want to do it if this is something that you're strongly opposed to (I'm not quite sure if it's a strong opposition or a case of I'm too busy and it's not important to me).

The alternatives that I've considered are:
1. I end up doing accounts in both Rev/Expense with matching names and possibly use account code to help me externally figure out whether to treat the pair of accounts as expense/revenue and writing custom P&L based on this... 
2. I investigated is directly editing the database to repoint the transaction from the revenue version of the account to the expense version (this actually worked really easily, but at the same time I think it's a bad idea for me to be "hacking" the database in this way.

Both of these feel like hacks and a lot of work, so my preference is to do, what seems like a simple change to allow negative values on deposit/withdrawal transactions which I expect would end up with result I'm looking for (as suggested above).



I've been using gnucash for years, and found firefly iii when recently research to work with a preferred modern web-based model and the API you've provided is very nice, so I'm excited about migrating!

**Comment by qskwood** [comment]
The TL;DR here is that the requests I made are not going to be implemented in the immediate term. The author uses the phrase "financial accounting" colloquially not literally (based on minimal research into accounting), doesn't want to make it behave like financial accounting bookkeeping software does, and thinks the software is fine as-is. If anyone has suggestions for improving Firefly III please open your own issues outlining what you'd like to get done, or fork the software and implement your changes locally.

N.B. to follow up on comments about why other people found Firefly III, I found it while researching alternatives to a paid app I was using after about a decade of the publisher ignoring workflow issues and raising prices. When the prices got to twice what I originally paid for it, I sent them an email asking why they were seemingly doing nothing to improve the software while jacking up the price. They were openly hostile and asserted that nobody wanted what I was asking for, that many of the features (that I relied on) were barely used and wouldn't get any investment from their team if they remained in the software at all, and that they were actually working super hard (they even showed me a changelog where for a two month period the only entry was adding images to their blog...) There is no reason, ever, to fight with a software publisher (whether they charge you money or not) when their goals for the software do not align with the work you want to get done. Just go use something else.
